File: Graph/GFG/Medium/Dijkstra/Dijkstra.cpp
// https://www.geeksforgeeks.org/problems/implementing-dijkstra-set-1-adjacency-matrix/1
#include<bits/stdc++.h>
using namespace std;
// USING PRIORITY QUEUE
vector<int> dijkstra(int V, vector<vector<int>> adj[], int S)
{
    // Code here
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

    pq.push({0, S});

    vector<int> minDist(V, INT_MAX);
    minDist[S] = 0;

    while (!pq.empty())
    {
        pair<int, int> curr = pq.top();
        int node = curr.second, dist = curr.first;
        pq.pop();

        for (auto &neighbor : adj[node])
        {
            int newDist = neighbor[1] + dist;
            if (minDist[neighbor[0]] > newDist)
            {
                minDist[neighbor[0]] = newDist;
                pq.push({newDist, neighbor[0]});
            }
        }
    }
    return minDist;
}
//USING SET
vector<int> dijkstra(int V, vector<vector<int>> adj[], int S)
{
    // Code here
    set<pair<int, int>> st;
    vector<int> dist(V, INT_MAX);

    dist[S] = 0;

    st.emplace(make_pair(0, S));

    while (!st.empty())
    {
        auto curr = *(st.begin());
        int node = curr.second;
        int distance = curr.first;

        st.erase(curr);

        for (auto neighbor : adj[node])
        {
            int newDist = neighbor[1] + distance;

            if (dist[neighbor[0]] > newDist)
            {
                if (dist[neighbor[0]] != INT_MAX)
                {
                    st.erase(make_pair(dist[neighbor[0]], neighbor[0]));
                }
                dist[neighbor[0]] = newDist;
                st.emplace(make_pair(newDist, neighbor[0]));
            }
        }
    }
    return dist;
}
int main(void) {
    
    return 0;
}

File: Graph/GFG/Medium/Dijkstra/Dijkstra.java
//https://www.geeksforgeeks.org/problems/implementing-dijkstra-set-1-adjacency-matrix/1

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.TreeSet;
class Pair {

    int distance, node;

    Pair(int distance, int node) {
        this.distance = distance;
        this.node = node;
    }

    // FOR TREESET
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        Pair pair = (Pair) obj;
        return pair.distance == distance && pair.node == node;
    }

    @Override
    public int hashCode() {
        return Objects.hash(distance, node);
    }
}
public class Dijkstra {
    // USING PRIORITY QUEUE
    static int[] dijkstra(int V, ArrayList<ArrayList<ArrayList<Integer>>> adj, int S) {
        // Write your code here
        Queue<Pair> pq = new PriorityQueue<>((x, y) -> {
            if (x.distance == y.distance) {
                return x.node - y.node;
            }
            return x.distance - y.distance;
        });

        pq.offer(new Pair(0, S));

        int[] minDist = new int[V];
        Arrays.fill(minDist, Integer.MAX_VALUE);

        minDist[S] = 0;

        while (!pq.isEmpty()) {
            Pair curr = pq.poll();

            int node = curr.node, distance = curr.distance;

            for (ArrayList<Integer> neighbor : adj.get(node)) {
                int newDist = distance + neighbor.get(1);

                if (minDist[neighbor.get(0)] > newDist) {
                    minDist[neighbor.get(0)] = newDist;
                    pq.offer(new Pair(newDist, neighbor.get(0)));
                }
            }
        }

        return minDist;
    }
    // USING TREESET
    static int[] dijkstra2(int V, ArrayList<ArrayList<ArrayList<Integer>>> adj, int S)
    {
        // Write your code here
        TreeSet<Pair> st = new TreeSet<>((x, y) -> {
            if(x.distance == y.distance) {
                return x.node - y.node;
            }
            return x.distance - y.distance;
        });
        
        st.add(new Pair(0, S));
        
        int[] minDist = new int[V];
        Arrays.fill(minDist, Integer.MAX_VALUE);
        
        minDist[S] = 0;
        
        while(!st.isEmpty()) {
            Pair curr = st.first();
            int node = curr.node, dist = curr.distance;
            st.remove(curr);
            for(ArrayList<Integer> neighbor: adj.get(node)) {
                int newDist = neighbor.get(1) + dist;
                
                if(minDist[neighbor.get(0)] > newDist) {
                    if(minDist[neighbor.get(0)] != Integer.MAX_VALUE) {
                        st.remove(new Pair(minDist[neighbor.get(0)], neighbor.get(0)));
                    }
                    minDist[neighbor.get(0)] = newDist;
                    st.add(new Pair(newDist, neighbor.get(0)));
                }
            }
        }
        return minDist;
    }
}
